File: oop/03-polymorphism/csharp/ChequeAccount.cs
/*
 * Demonstrates: Polymorphism - Override
 * Concept: ChequeAccount overrides ProcessWithdrawal to allow overdraft
 */

namespace FinanceDomain.Polymorphism;

public class ChequeAccount : Account
{
    public decimal OverdraftLimit { get; }

    public ChequeAccount(string accountNumber, string accountHolder, decimal openingBalance, decimal overdraftLimit)
        : base(accountNumber, accountHolder, openingBalance)
    {
        OverdraftLimit = overdraftLimit;
    }

    public override TransactionResult ProcessWithdrawal(decimal amount)
    {
        if (amount <= 0)
            return TransactionResult.Failed("Amount must be greater than zero.");

        decimal availableFunds = _balance + OverdraftLimit;

        if (amount > availableFunds)
            return TransactionResult.Failed($"Exceeds overdraft limit. Available: R{availableFunds:F2}");

        _balance -= amount;
        return TransactionResult.Succeeded($"Withdrew R{amount:F2}. New balance: R{_balance:F2}");
    }

    public override string GetAccountType() => "Cheque Account";
}

File: oop/03-polymorphism/typescript/chequeAccount.ts
/*
 * Demonstrates: Polymorphism - Override
 * Concept: ChequeAccount overrides processWithdrawal to allow overdraft
 */

import { Account, TransactionResult } from "./account";

export class ChequeAccount extends Account {
    public readonly overdraftLimit: number;

    constructor(accountNumber: string, accountHolder: string, openingBalance: number, overdraftLimit: number) {
        super(accountNumber, accountHolder, openingBalance);
        this.overdraftLimit = overdraftLimit;
    }

    override processWithdrawal(amount: number): TransactionResult {
        if (amount <= 0) {
            return TransactionResult.failed("Amount must be greater than zero.");
        }

        const availableFunds = this._balance + this.overdraftLimit;

        if (amount > availableFunds) {
            return TransactionResult.failed(`Exceeds overdraft limit. Available: R${availableFunds.toFixed(2)}`);
        }

        this._balance -= amount;
        return TransactionResult.succeeded(`Withdrew R${amount.toFixed(2)}. New balance: R${this._balance.toFixed(2)}`);
    }

    override getAccountType(): string {
        return "Cheque Account";
    }
}
